rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+
+Note: this mode is a stretch goal and is not scheduled to be completed, but
+this RFD would be incomplete if it did not at least consider this future work.
+
+### Client interface
+
+In order to utilize the Teleport Key Agent, dependent clients will need to
+interface with the agent as opposed to file storage or in-memory storage. Agent
+storage will be added as a new type of storage, with a `Keystore` implementation
+for [signing mode](#signing-mode) and a `ProfileStore` and `TrustedCertsStore`
+implementation for [login sync mode](#login-sync-mode-future-work).
+
+### Security
+
+#### Local key agent
+
+For the intended use case of using Teleport Key Agent as a local key agent,
+there is not much of concern to consider:
+
+* The Teleport Key Agent acts as a proxy for normal file storage, and as such
+is protected through file permissions on the unix socket it's served on:
+`$TELEPORT_HOME/agent.sock`.
+* Even in login sync mode, the agent never exposes the underlying private key
+material, only an interface to sign with.
+
+#### Key agent forwarding
+
+We must also consider the unintended use case of forwarding the agent over ssh
+with unix domain socket forwarding. Like ssh agent forwarding, this is an
+insecure use case which will be strongly advised against. However, it cannot
+be entirely avoided, the same way that we can not stop a user from running
+`tsh scp $HOME/.tsh server01:`.
+
+As usual remote attacks are best mitigated through phish-proof authentication.
+If the following features are enabled, the impact of a compromised forwarded
+Teleport Key Agent will be largely mitigated:
+
+* Per-session MFA or Hardware Key Support (pin/touch)
+* MFA for Admin Actions
+
+If deemed necessary, we could put the agent behind an extra
+[confirmation layer](#confirmation-prompt-future-work). This confirmation layer
+would require dependent clients to perform a confirmation action (y/N, pin, or
+touch) to allow the dependent client to utilize the agent.
+
+### Configuration
+
+See [running the agent](#running-the-agent) for how to configure Teleport
+Connect to run the Teleport Key Agent.
+
+### Proto specification
+
+```proto
+// KeyAgentService provides a Teleport key agent service, allowing multiple Teleport client
+// processes to share a single instance of a Teleport key. This is useful for PIV keys which
+// can only be accessed by one process at a time.
+service KeyAgentService {
+  // Query checks whether the provided key index has an existing keyring and
+  // returns the public keys.
+  rpc Query(QueryRequest) returns (QueryResponse) {}
+  // Sign signs the given digest with the private key corresponding to the given
+  // key index and type. If a hash or salt was used to produce the digest, HashName
+  // and SaltLength must be provided as well.
+  //
+  // This rpc can be used to implement Go's crypto.Signer interface.
+  rpc Sign(SignRequest) returns (SignResponse) {}
```

Comment:
This is very flexible as a programming interface, but it seems overly generous in security terms. Should we design the agent to be more purpose-specific, so its uses are more focused and less prone to unintended use?

Reply:
This is meant to solve the same problem as `ssh-agent` and `gpg-agent`, providing a local API for signing with local private keys. As such the implementation is similar. What unintended uses are we are concerned about exactly?

Reply:
Hmm I see what you mean. Basically we would need to make `tsh` work with `tshd` and expand `tshd` to cover any action which we want to support PIN caching for.

Reply:
I've added a `tshd` section in additional considerations.